framework/Optimizers/GradientBasedOptimizer.py

```diff
@@ -187,6 +187,21 @@ def localGenerateInput(self,model,oldInput):
     """
     self.readyVarsUpdate = {traj:False for traj in self.optTrajLive}
 
+  def calculateMultivectorMagnitude(self,values):
+    """
+      Calculates the magnitude of vector "values", where values might be a combination of scalars and vectors (but not matrices [yet]).
+      Calculates the magnitude as if "values" were flattened into a 1d array.
+      @ In, values, list, values for which the magnitude will be calculated
+      @ Out, mag, float, magnitude
+    """
+    # use np.linalg.norm (Frobenius norm) to calculate magnitude
+    ## pre-normalise vectors, this is mathematically equivalent to flattening the vector first
+    ## NOTE this assumes gradient vectors are 0 or 1 dimensional, not 2 or more! (vectors or scalars, not matrices)
+    # TODO this could be sped up if we could avoid calling np.atleast_1d twice, but net slower if we loop first
+    preMag = [np.linalg.norm(val) if len(np.atleast_1d(val))>1 else np.atleast_1d(val)[0] for val in values]
```

Comment:
The type and shape of these variables matters a lot passing through here.  I had to add that term specifically because it was failing to do what it is supposed to do in the event a 1-length np.ndarray is val, but often receives floats as well.